File: tests/algorithms/transpiler/FS/graphics/bezier_curve.fs
// Generated 2025-08-13 16:13 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    let s = if s.EndsWith(".0") then s.Substring(0, s.Length - 2) else s
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec n_choose_k (n: int) (k: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable n = n
    let mutable k = k
    try
        if (k < 0) || (k > n) then
            __ret <- 0.0
            raise Return
        if (k = 0) || (k = n) then
            __ret <- 1.0
            raise Return
        let mutable result: float = 1.0
        let mutable i: int = 1
        while i <= k do
            result <- (result * (1.0 * (float ((n - k) + i)))) / (1.0 * (float i))
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and pow_float (``base``: float) (exp: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: float = 1.0
        let mutable i: int = 0
        while i < exp do
            result <- result * ``base``
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and basis_function (points: float array array) (t: float) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable points = points
    let mutable t = t
    try
        let degree: int = (Seq.length (points)) - 1
        let mutable res: float array = Array.empty<float>
        let mutable i: int = 0
        while i <= degree do
            let coef: float = n_choose_k (degree) (i)
            let term: float = (pow_float (1.0 - t) (degree - i)) * (pow_float (t) (i))
            res <- Array.append res [|(coef * term)|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and bezier_point (points: float array array) (t: float) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable points = points
    let mutable t = t
    try
        let basis: float array = basis_function (points) (t)
        let mutable x: float = 0.0
        let mutable y: float = 0.0
        let mutable i: int = 0
        while i < (Seq.length (points)) do
            x <- x + ((_idx basis (int i)) * (_idx (_idx points (int i)) (int 0)))
            y <- y + ((_idx basis (int i)) * (_idx (_idx points (int i)) (int 1)))
            i <- i + 1
        __ret <- unbox<float array> [|x; y|]
        raise Return
        __ret
    with
        | Return -> __ret
let control: float array array = [|[|1.0; 1.0|]; [|1.0; 2.0|]|]
ignore (printfn "%s" (_str (basis_function (control) (0.0))))
ignore (printfn "%s" (_str (basis_function (control) (1.0))))
ignore (printfn "%s" (_str (bezier_point (control) (0.0))))
ignore (printfn "%s" (_str (bezier_point (control) (1.0))))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/graphics/bezier_curve.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function n_choose_k($n, $k) {
  global $control;
  if ($k < 0 || $k > $n) {
  return 0.0;
}
  if ($k == 0 || $k == $n) {
  return 1.0;
}
  $result = 1.0;
  $i = 1;
  while ($i <= $k) {
  $result = $result * (1.0 * ($n - $k + $i)) / (1.0 * $i);
  $i = $i + 1;
};
  return $result;
};
  function pow_float($base, $exp) {
  global $control;
  $result = 1.0;
  $i = 0;
  while ($i < $exp) {
  $result = $result * $base;
  $i = $i + 1;
};
  return $result;
};
  function basis_function($points, $t) {
  global $control;
  $degree = count($points) - 1;
  $res = [];
  $i = 0;
  while ($i <= $degree) {
  $coef = n_choose_k($degree, $i);
  $term = pow_float(1.0 - $t, $degree - $i) * pow_float($t, $i);
  $res = _append($res, $coef * $term);
  $i = $i + 1;
};
  return $res;
};
  function bezier_point($points, $t) {
  global $control;
  $basis = basis_function($points, $t);
  $x = 0.0;
  $y = 0.0;
  $i = 0;
  while ($i < count($points)) {
  $x = $x + $basis[$i] * $points[$i][0];
  $y = $y + $basis[$i] * $points[$i][1];
  $i = $i + 1;
};
  return [$x, $y];
};
  $control = [[1.0, 1.0], [1.0, 2.0]];
  echo rtrim(_str(basis_function($control, 0.0))), PHP_EOL;
  echo rtrim(_str(basis_function($control, 1.0))), PHP_EOL;
  echo rtrim(_str(bezier_point($control, 0.0))), PHP_EOL;
  echo rtrim(_str(bezier_point($control, 1.0))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
